lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java

```diff
@@ -101,12 +115,43 @@ private Class getPropAsClass(Schema schema, String prop) {
     if (name == null)
       return null;
     try {
-      return ClassUtils.forName(getData().getClassLoader(), name);
+      Class clazz = ClassUtils.forName(getData().getClassLoader(), name);
+      checkSecurity(clazz);
+      return clazz;
     } catch (ClassNotFoundException e) {
       throw new AvroRuntimeException(e);
     }
   }
 
+  private boolean trustAllPackages() {
+    return (trustedPackages.size() == 1 && "*".equals(trustedPackages.get(0)));
+  }
+
+  private void checkSecurity(Class clazz) throws ClassNotFoundException {
+    if (trustAllPackages() || clazz.isPrimitive()) {
+      return;
+    }
+
+    boolean found = false;
+    Package thePackage = clazz.getPackage();
+    if (thePackage != null) {
```

Comment:
@Fokko or @martin-g 
When a class is not in any package `clazz.getPackage()` returns null and this condition renders that class trusty.

Move `if (!found) throw new SecurityException` out of the loop.

Reply:
I can take a look on this too as I'm the original author

Reply:
@MichalFoksa Would you like to send a Pull Request?

Reply:
Else I can work on a PR tomorrow

Reply:
@martin-g  Here you are PR #3311